src/compiler/scanner.ts

```diff
@@ -2052,12 +2052,38 @@ namespace ts {
                             return token = SyntaxKind.Unknown;
                         }
 
-                        if (isIdentifierStart(codePointAt(text, pos + 1), languageVersion)) {
+                        const charAfterHash = codePointAt(text, pos + 1);
+                        if (charAfterHash === CharacterCodes.backslash) {
                             pos++;
-                            scanIdentifier(codePointAt(text, pos), languageVersion);
+                            const extendedCookedChar = peekExtendedUnicodeEscape();
+                            if (extendedCookedChar >= 0 && isIdentifierStart(extendedCookedChar, languageVersion)) {
+                                pos += 3;
+                                tokenFlags |= TokenFlags.ExtendedUnicodeEscape;
+                                tokenValue = "#" + scanExtendedUnicodeEscape() + scanIdentifierParts();
+                                return token = SyntaxKind.PrivateIdentifier;
+                            }
+
+                            const cookedChar = peekUnicodeEscape();
+                            if (cookedChar >= 0 && isIdentifierStart(cookedChar, languageVersion)) {
+                                pos += 6;
+                                tokenFlags |= TokenFlags.UnicodeEscape;
+                                tokenValue = "#" + String.fromCharCode(cookedChar) + scanIdentifierParts();
+                                return token = SyntaxKind.PrivateIdentifier;
+                            }
+                            pos--;
+                        }
+
+                        if (isIdentifierStart(charAfterHash, languageVersion)) {
+                            pos++;
+                            // We're relying on scanIdentifier's behavior and adjusting the token kind after the fact.
+                            // Notably absent from this block is the fact that calling a function named "scanIdentifier",
+                            // but identifiers don't include '#', and that function doesn't deal with it at all.
+                            // This works because 'scanIdentifier' tries to reuse source characters and builds up substrings;
+                            // however, it starts at the 'tokenPos' which includes the '#', and will "accidentally" prepend the '#' for us.
+                            scanIdentifier(charAfterHash, languageVersion);

```

Comment:
This change looks consistent with unicode escape handling elsewhere in the scanner, but I'm not sure I understand why `scanIdentifier` doesn't handle them.  Is it illegal to start an identifier with an escape sequence?  (Maybe that would introduce an ambiguity, but none jumps to mind.)

Reply:
Am I reading [this](https://262.ecma-international.org/13.0/#prod-IdentifierStart) correctly? It looks like any identifier _can_ start with a unicode escape.  If so, why wouldn't the fix be in `scanIdentifier`?

Reply:
@DanielRosenwasser Last question ⬆️

Reply:
`scanIdentifier` could handle them if we just checked if the first character was either an identifier start or a `\`, but right now the logic is to not advance if we can't get at least one complete identifier start.

Right now the handling for any of the following incomplete escape sequences...

```
\
\u
\u0
\u00
\u000
\u{}
```

is to not munch up these characters, and identify the `\` as an unknown token, followed by whatever.

As an extension, what the current code does with private fields is to make each of

```
#\
#\u
#\u0
#\u00
#\u000
#\u{}
```

an incomplete private field with the name `#`, followed by an unknown token `\` followed by whatever.

Arguably, private fields could diverge here for some better errors.